tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_runner.cc

```diff
@@ -0,0 +1,299 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_runner.h"
+
+#include "absl/strings/str_cat.h"
+#include "tensorflow/compiler/xla/layout_util.h"
+#include "tensorflow/compiler/xla/service/gpu/backend_configs.pb.h"
+#include "tensorflow/compiler/xla/service/gpu/stream_executor_util.h"
+#include "tensorflow/compiler/xla/shape_util.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+#include "tensorflow/compiler/xla/util.h"
+
+namespace xla {
+namespace gpu {
+namespace {
+
+struct CudnnBatchNormParamsCommon {
+  se::DeviceMemoryBase operand;
+  se::dnn::BatchDescriptor operand_desc;
+  se::dnn::BatchDescriptor scale_offset_desc;
+  se::DeviceMemory<float> scale;
+  float epsilon;
+};
+
+struct CudnnBatchNormForwardInferenceParams {
+  CudnnBatchNormParamsCommon common;
+  se::DeviceMemoryBase output;
+  se::DeviceMemory<float> offset;
+  se::DeviceMemory<float> mean;
+  se::DeviceMemory<float> variance;
+};
+
+struct CudnnBatchNormForwardTrainingParams {
+  CudnnBatchNormParamsCommon common;
+  se::DeviceMemoryBase output_data;
+  se::DeviceMemory<float> offset;
+  se::DeviceMemory<float> output_mean;
+  se::DeviceMemory<float> output_inv_stddev;
+};
+
+struct CudnnBatchNormBackwardParams {
+  CudnnBatchNormParamsCommon common;
+  se::DeviceMemoryBase output_grad_data;
+  se::DeviceMemoryBase grad_output;
+  se::DeviceMemory<float> output_grad_scale;
+  se::DeviceMemory<float> output_grad_offset;
+  se::DeviceMemory<float> mean;
+  se::DeviceMemory<float> inv_stddev;
+};
+
+std::pair<se::dnn::BatchDescriptor /*input_desc*/,
+          se::dnn::BatchDescriptor /*scale_offset_desc*/>
```

Comment:
This should probably just be a struct since the two fields have the same type and thus can be accidentally switched.

Reply:
Done